### DISCUSSION TITLE
CIP 0030 wallets and pycardano

### DISCUSSION THREAD

**Discussion by bapshago** [Python-Cardano/pycardano#173] [General]
Is anyone using a CIP 0030 enabled browser wallet to sign py cardano transactions?  

When I send just a txbody the wallet will not read the transaction.  When I send a fully signed tx the wallet will sign the tx and send back the witness set.  When I try to combine the witness set with the original body of the transaction blockfrost or the wallet will not submit the transaction.

Here is the path I am using:
1) Obtain a used address from wallet API. (api.getUsedAddresses)
2) Build an address in pycardano using the obtained address with sufficient funds.
3) Validate the address is not delegated to the pool selected for delegation and determine if address needs to be registered.
4) Use transaction builder to build the body of transaction for pool delegation using the address.
5) Build and Sign the TX using a **non related** stake and payment signing key.
6) Submit the tx to the wallet to sign. 
7) Sign the tx with the wallet (transaction appears correct in the wallets)
8) Use the returned witness set and combine it with the original tx body to complete the "txbuild and sign" with the correct address.
9) Try to submit the tx either via blockfrost or via the wallet api.

I would like to skip step 5 but I can not get a wallet to accept a txbody generated by pycardano.   Has anyone done this yet?

**Comment by bapshago** [comment]
I was able to work around this problem by building up the transaction manually, not using the tx builder. The path was as follows:

1) Obtain a used address from wallet API. (api.getUsedAddresses)
2) Using the address, look up available utxos on blockfrost with sufficient funds.
3) Use that utxo to build a transaction.
4) Add required outputs for the tx.
5) Build and Sign the TX using an empty transaction witness set. transwit = TransactionWitnessSet()
6) Submit the tx to the wallet to sign.
7) Sign the tx with the wallet .
8) Use the returned witness set from the wallet and combine it with the original tx body complete the txbuild. signed_tx = T) ransaction(tx_body, transwitset)
9) Submit the tx via blockfrost.

The transaction builder is nice. It would be nice if it allowed you to sign a transaction witness set with empty data, because the CIP 0030 wallets are expecting the tx to include it.

**Reply by cffls** [reply]
Apologies for being a bit late on this question. Looks like you are trying to construct a tx with empty witness. Have you tried passing an empty list to build_and_sign like this `tx = txbuilder.build_and_sign(signing_keys=[], ...)`?
